### PR TITLE
[worker] fix vLLM multimodal encoder_cache GPU memory leak

### PR DESCRIPTION
**Pull request by raja-kumar**
### Problem

When training multimodal models (e.g. Qwen3-VL) with the colocated HybridEngine, GPU memory grows roughly linearly across training steps and eventually OOMs in the actor update (`update_actor` → `loss.backward()`). The leak is **image-only** (text-only runs are stable) and **resets on restart**, so resuming from a checkpoint runs for a bounded number of steps and then OOMs again.

### Root cause

vLLM's V1 `GpuModelRunner` keeps an `encoder_cache` (`dict[mm_hash -> GPU tensor]` of image/video encoder outputs). Entries are only evicted by the scheduler's `EncoderCacheManager` **under capacity pressure** — it frees the oldest freeable entries only when a new request needs slots.

In the colocated HybridEngine setup, each training step runs a single bounded `generate()` call and then sleeps, so that pressure never builds: each step's batch fits within the cache budget, the "need more slots" eviction path never fires, and finished-request entries are never reclaimed. The cache therefore grows by ~one entry per image per step (~0.25 GB/step for Qwen3‑VL on geometry3k) until CUDA OOM.

Two details make it stick:
- `sleep(level=1)` offloads only model **weights** (via `CuMemAllocator` with `offload_tags=("weights",)`), not this cache.
- `torch.cuda.empty_cache()` cannot reclaim it because the entries are still referenced by the dict.

Because vLLM runs in-process (SPMD / `external_launcher`) and shares the process CUDA allocator with training, the growth shows up as `torch.cuda.memory_allocated()`.

### Fix

Drain the encoder cache in `FSDPVLLMShardingManager.offload_vllm()` right after `sleep()`. Generation has fully completed by then (no unfinished requests), so it is safe to clear the runner's cached tensors and reset the scheduler-side `EncoderCacheManager` slot accounting to its empty state. The reset is best-effort and fully guarded (`try/except`, attribute checks, `has_unfinished_requests()`), so any vLLM-internal layout change degrades to a no-op rather than breaking training.

### Verification

Qwen3-VL-8B + geometry3k, colocated HybridEngine:
- `encoder_cache` stays at **0 entries per step** (was growing ~32/step).
- Post-rollout GPU usage goes **flat** instead of climbing ~0.25 GB/step; the OOM no longer occurs.
- **Step-1 reward/accuracy is bit-identical** to before the change (the drain runs after generation and does not touch the compute path); later steps differ only by normal RL run-to-run noise.

Relates to the memory-leak reports in #444 (and similar: #469, #398).

### CODE REVIEW

File: verl/workers/sharding_manager/fsdp_vllm.py [hunk 1]
```diff
@@ -207,13 +207,59 @@ def offload_vllm(self):
...
+            manager = getattr(scheduler, "encoder_cache_manager", None)
+            if manager is not None and not scheduler.has_unfinished_requests():
+                manager.num_free_slots = manager.cache_size
+                manager.num_freeable_slots = manager.cache_size
+                manager.cached.clear()
+                if hasattr(manager, "freeable"):
+                    manager.freeable.clear()
+                manager.freed.clear()
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (raja-kumar) [reply verl/workers/sharding_manager/fsdp_vllm.py:261]: Thanks for the review. I don't think this specific change is needed: `getattr(obj, name, default)` returns `default` when `obj` is `None` — it does **not** raise `AttributeError` (that only happens with attribute access or `getattr` *without* a default):

```python
>>> getattr(None, "scheduler", None)
None   # no exception
```

So in the current code, if `engine_core` or `scheduler` is `None`, the chained `getattr(..., None)` calls just yield `None` and `manager` ends up `None`, which the `if manager is not None and not scheduler.has_unfinished_requests():` guard handles (and `and` short-circuits before `scheduler.has_unfinished_requests()` is ever reached, since `manager` can only be non-`None` when `scheduler` is non-`None`).

The whole block is also wrapped in `try/except Exception: pass` specifically so that any unexpected vLLM-internal layout change degrades to a no-op rather than breaking training. Happy to switch to explicit `is not None` checks if you'd prefer the more defensive style for readability, but functionally the current version is already safe.

**Review by hiyouga — APPROVED** [review]
LGTM
